File: Java/src/com/daily/algothrim/linked/algo/Reversal.java
package com.daily.algothrim.linked.algo;

import com.daily.algothrim.linked.LinkedNode;

/**
 * 单链表反转
 */
public class Reversal {

    public static void main(String[] args) {
        new Reversal().reversal(new LinkedNode<>("a",
                new LinkedNode<>("b",
                        new LinkedNode<>("c",
                                new LinkedNode<>("d",
                                        new LinkedNode<>("a", null)))))).printAll();
    }

    /**
     * O(n)
     */
    private LinkedNode<String> reversal(LinkedNode<String> singleLinked) {
        LinkedNode<String> result = null;
        LinkedNode<String> temp;

        while (singleLinked != null) {
            temp = singleLinked.next;
            singleLinked.next = result;
            result = singleLinked;
            singleLinked = temp;
        }

        return result;
    }
}

File: kotlin/src/com/daily/algothrim/linked/algo/Reversal.kt
package com.daily.algothrim.linked.algo

import com.daily.algothrim.linked.LinkedNode

/**
 * 单链表反转
 */
class Reversal {

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            Reversal().reversal(LinkedNode("a").apply {
                next = LinkedNode("b").apply {
                    next = LinkedNode("c").apply {
                        next = LinkedNode("d").apply {
                            next = LinkedNode("a")
                        }
                    }
                }
            })?.printAll()
        }
    }

    /**
     * O(n)
     */
    fun reversal(singleLinked: LinkedNode<String>?): LinkedNode<String>? {
        var result: LinkedNode<String>? = null
        var current = singleLinked
        var next: LinkedNode<String>?

        while (current != null) {
            next = current.next
            current.next = result
            result = current
            current = next
        }
        return result
    }
}